core/src/impl/Kokkos_Core.cpp

```diff
@@ -77,6 +78,54 @@ std::stack<hook_function_type, std::list<hook_function_type>> finalize_hooks;
 namespace Kokkos {
 namespace Impl {
 
+ExecSpaceManager ExecSpaceManager::g_space_initializer;
+
+ExecSpaceManager& ExecSpaceManager::get_instance() {
+  return g_space_initializer;
+}
+
+void ExecSpaceManager::register_space_factory(
+    const std::string name, std::unique_ptr<ExecSpaceInitializerBase> space) {
+  printf("registering exec space: %s \n", name.c_str());
+  exec_space_factory_list[name] = std::move(space);
+}
+
+void ExecSpaceManager::initialize_spaces(const Kokkos::InitArguments& args) {
+  for (std::map<std::string,
+                std::unique_ptr<ExecSpaceInitializerBase>>::iterator it =
+           exec_space_factory_list.begin();
+       it != exec_space_factory_list.end(); it++) {
+    it->second->initialize(args);
+  }
```

Comment:
Please modernize (here and elsewhere, unless this is moved from somewhere)

```suggestion
  for (auto& to_init : exec_space_factory_list) {
    to_init.second->initialize(args);
  }
```

Reply:
resolved with 424fc647